### PR TITLE
 Integrate CI for Running Benchmarks #436 

### LINKED ISSUE #436: Integrate CI for Running Benchmarks
### Overview

We need to enhance our CI pipeline to not only run our existing Criterion benchmarks but also ensure that the build time stays within acceptable limits. Moreover, a clear and concise report should be generated and published on the PR for quick and easy understanding.

### Requirements

1. **Run Criterion Benchmarks:** Our CI should execute the existing Criterion benchmarks as part of the pipeline.
2. **Fail on Degradation:** The CI should fail if there's a performance degradation when compared to the `main` branch.
3. ~**Build Time Check:** The CI should compare the build time of the current PR with the average build time of the last few successful builds on the `main` branch. If the build time increases by more than 10%, the CI should fail.~
4. **Benchmark Report:** A tabular comparison of benchmarks should be published on the PR, similar to the code-cov report. It should display the current benchmark results alongside the results from the `main` branch for easy comparison.
5. **Conditional Execution:** Benchmarks and build time checks should only be executed when the `benchmark` label is added to a PR or commit.
6. **Support for Forks:** The CI process should be designed to run benchmarks on forks as well.

### Rationale

Performance is paramount for our project. It's essential to ensure that any changes do not adversely affect either the runtime performance (tracked by benchmarks) or the build time. By integrating these checks into our CI, we can maintain a high standard of performance and ensure that contributors receive immediate feedback on any potential issues.

### PR DESCRIPTION
**Pull request by alankritdabral**
**Summary**:
I have used **Criterion**  to check performance degradation.

To compare Criterion benchmarks  i have use  https://github.com/BurntSushi/critcmp

Additions  made :

benches/iai-callgrind
src/benchmark (changes in the benchmarks after addding **src/benchmarks** [iai-callgrind.txt](https://github.com/tailcallhq/tailcall/files/13473598/iai-callgrind.txt) [criterion.txt](https://github.com/tailcallhq/tailcall/files/13473607/criterion.txt) run on local device)

Issue Reference(s):
Fixes #436 

- [x] I ran `cargo test` successfully.
- [x] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [x] I have added relevant unit & integration tests.
- [x] I have updated the documentation accordingly (if applicable).
- [x] I have performed a self-review of my own code.

/claim #436 
[criterion.txt](https://github.com/tailcallhq/tailcall/files/13473602/criterion.txt)

### CODE REVIEW

**Comment by alankritdabral** [comment]
@tusharmath

**Comment by AlooDon** [comment]
@tusharmath  can you review it and tell if anything is not right!!

**Comment by tusharmath** [comment]
Hey @alankritdabral ! Thanks for the PR, can you suggest a way to test ?

**Comment by alankritdabral** [comment]
> Hey @alankritdabral ! Thanks for the PR, can you suggest a way to test ?

we can create a dummy issue with **benchmark** label  and it will run the benchmarks 
This can be one way

**Comment by alankritdabral** [comment]
@tusharmath asi have said earlier https://github.com/alankritdabral/tailcall/pull/34#issue-2008631651 this is the example

File: .github/scripts/compare_build_time.sh [hunk 1]
```diff
@@ -0,0 +1,106 @@
+#!/bin/bash
```
Comment (meskill) [inline .github/scripts/compare_build_time.sh:1]: this file is more like iai-callgrind benchmarks and not compare_build_time as it runs benchmarks for the code right?

for comparing build time we need to measure how long is the compilation of the project

File: .github/scripts/compare_build_time.sh [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+                fail_ci=1
+            fi
+        else
+            total_read_write=$((l1_hits + l2_hits + ram_hits))
+            estimated_cycles=$((l1_hits + 5 * l2_hits + 35 * ram_hits))
+        fi
+
+        x=$((x + total_read_write))
+        y=$((y + estimated_cycles))
+    done
```
Comment (meskill) [inline .github/scripts/compare_build_time.sh:101]: for previous pr @tusharmath has suggested to use `cargo-script` for such kind of scripts.

wdyt about it? simple bash scripts are ok overall but when they get more complex it's really a burden to support it. Or consider another script language that is more reader friendly

File: .github/workflows/benchmarks.yml [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+            })
+              ).data[0].number;
+            }
+          result-encoding: string
+          
+      - name: Comment on PR
+        uses: actions/github-script@v6
+        with:
+          github-token: ${{ secrets.GITHUB_TOKEN }}
+          script: |
```
Comment (meskill) [inline .github/workflows/benchmarks.yml:73]: should be made as separate file to include here

File: .github/workflows/benchmarks.yml [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+      - name: Install iai-callgrind-runner
+        run: cargo install --version 0.7.3 iai-callgrind-runner
+
+      - name: Install Valgrind
+        run: sudo apt-get update && sudo apt-get install valgrind
+
+      - name: Execute build_time_compare.sh
+        run: |
+          chmod +x .github/scripts/compare_build_time.sh
```
Comment (meskill) [inline .github/workflows/benchmarks.yml:113]: just adding execute flags to script and committing it should work without additional steps in ci

File: .github/workflows/benchmarks.yml [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+      - name: Install Critcmp
+        run: cargo install critcmp 
+
+      - name: Benchmark changes
+        run: |
+          cargo bench --bench impl_path_string_for_evaluation_context -- --save-baseline new_branch
+          cargo bench --bench data_loader_bench -- --save-baseline new_branch
+          cargo bench --bench request_template_bench -- --save-baseline new_branch
+          cargo bench --bench json_like_bench -- --save-baseline new_branch
```
Comment (meskill) [inline .github/workflows/benchmarks.yml:29]: do we need to run benchmarks here one-by-one? `cargo bench` run all the benchmarks by default

File: benches/iai-callgrind/impl_path_string_for_evaluation_context_iai-callgrind.rs [hunk 6]
```diff
@@ -0,0 +1,156 @@
+use std::borrow::Cow;
```
Comment (meskill) [inline benches/iai-callgrind/impl_path_string_for_evaluation_context_iai-callgrind.rs:1]: I've mentioned it in the previous pr and would like to point it again. It'll be amazing if we could have share the code between different benchmarks runners.

I believe it could be done by using cargo features and share as much code as possible

File: benches/iai-callgrind/impl_path_string_for_evaluation_context_iai-callgrind.rs [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    Some(Cow::Borrowed("header"))
+  );
+  assert_eq!(eval_ctx.path_string(&["headers", "x-missing"]), None);
+
+  // vars
+  assert_eq!(eval_ctx.path_string(&["vars", "existing"]), Some(Cow::Borrowed("var")));
+  assert_eq!(eval_ctx.path_string(&["vars", "missing"]), None);
+}
+
+#[library_benchmark]
```
Comment (meskill) [inline benches/iai-callgrind/impl_path_string_for_evaluation_context_iai-callgrind.rs:131]: turns out with iai we benchmark the setup code as well, but it seems like that is limitation from the iai itself: "Con: Iai cannot exclude setup code from the measurements, where Criterion-rs can"

**Review by meskill — CHANGES_REQUESTED** [review]

File: benches/iai-callgrind/impl_path_string_for_evaluation_context_iai-callgrind.rs (hunk 6 above)
Comment (AlooDon) [reply benches/iai-callgrind/impl_path_string_for_evaluation_context_iai-callgrind.rs:1]: can you please elaborate this

File: .github/scripts/compare_build_time.sh (hunk 1 above)
Comment (AlooDon) [reply .github/scripts/compare_build_time.sh:1]: made the changes accordingly

File: .github/scripts/compare_build_time.sh (hunk 2 above)
Comment (AlooDon) [reply .github/scripts/compare_build_time.sh]: i have shorten the code and it looks simple to me  
this code will always work in comparing iai-callgrind and will never need any changes

File: .github/workflows/benchmarks.yml (hunk 3 above)
Comment (AlooDon) [reply .github/workflows/benchmarks.yml]: merged this file with bot.yml

File: .github/scripts/compare_build_time.sh (hunk 2 above)
Comment (meskill) [reply .github/scripts/compare_build_time.sh]: well, the experience says that any code may require changes despite initial intentions for it

File: benches/iai-callgrind/impl_path_string_for_evaluation_context_iai-callgrind.rs (hunk 6 above)
Comment (meskill) [reply benches/iai-callgrind/impl_path_string_for_evaluation_context_iai-callgrind.rs:1]: could you please specify what is unclear for you?

I can give some suggestion on how to do this, but the exact way to solve it is on you.

File: .github/scripts/compare_build_time.sh (hunk 2 above)
Comment (alankritdabral) [reply .github/scripts/compare_build_time.sh]: ok, but does the current code looks simple to deal with or should use cargo script

File: .github/workflows/benchmarks.yml (hunk 5 above)
Comment (AlooDon) [reply .github/workflows/benchmarks.yml]: sloved using  cargo bench --features criterion -- --save-baseline main_branch

File: .github/workflows/benchmarks.yml (hunk 4 above)
Comment (AlooDon) [reply .github/workflows/benchmarks.yml]: solved

File: benches/iai-callgrind/impl_path_string_for_evaluation_context_iai-callgrind.rs (hunk 6 above)
Comment (AlooDon) [reply benches/iai-callgrind/impl_path_string_for_evaluation_context_iai-callgrind.rs:1]: yeah sure please i just need 1 example

File: benches/iai-callgrind/impl_path_string_for_evaluation_context_iai-callgrind.rs (hunk 7 above)
Comment (AlooDon) [reply benches/iai-callgrind/impl_path_string_for_evaluation_context_iai-callgrind.rs:131]: yes

File: benches/iai-callgrind/impl_path_string_for_evaluation_context_iai-callgrind.rs (hunk 6 above)
Comment (AlooDon) [reply benches/iai-callgrind/impl_path_string_for_evaluation_context_iai-callgrind.rs]: @meskill can you check the latest changes

**Comment by alankritdabral** [comment]
@tusharmath can you please increase the  bounty by $50

**Comment by amitksingh1490** [comment]
@alankritdabral The benchmark files have a lot of changes, Ideally, we should minimise the diff in those unless required. 
Cleanups can be done in separate PR ideally.

**Comment by alankritdabral** [comment]
> @alankritdabral The benchmark files have a lot of changes, Ideally, we should minimise the diff in those unless required. Cleanups can be done in separate PR

File: benches/iai-callgrind/impl_path_string_for_evaluation_context_iai-callgrind.rs (hunk 6 above)
Comment (meskill) [reply benches/iai-callgrind/impl_path_string_for_evaluation_context_iai-callgrind.rs:1]: I saw you had implemented something to share code between test runners. It didn't work eventually?

File: .github/scripts/check_degradation.sh [hunk 8]
```diff
@@ -0,0 +1,23 @@
+current_branch=$(git rev-parse --abbrev-ref HEAD) 
+critcmp new_branch main_branch | awk 'NR>2 {
+    item = $1
+    before = $3
+    after = $7
+    before_val = ($3 ~ /ns/) ? $3 : ($3 ~ /µs/) ? $3 * 1000 : ($3 ~ /ms/) ? $3 * 1000000 : "invalid"
+    after_val = ($7 ~ /ns/) ? $7 : ($7 ~ /µs/) ? $7 * 1000 : ($7 ~ /ms/) ? $7 * 1000000 : "invalid"
```
Comment (meskill) [inline .github/scripts/check_degradation.sh:7]: I think the critcmp is already showing the relative comparison between measurements

image

Can we use them instead of calculating it?

File: .github/scripts/check_degradation.sh [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+
+    if (before_val != "invalid" && after_val != "invalid") {
+        change = ((after_val - before_val) / before_val) * 100
+        gsub("%", "", change)  # Remove '%' symbol
+
+        printf "| %-30s | %-20s | %-20s | %-10.2f |\n", item, before, after, change >> "benches/critcmp.txt"
+
+        if (change > 10) {
+            printf "Percentage change for %s exceeds 10%%. Failing the workflow.\n", item
+            exit 1
```
Comment (meskill) [inline .github/scripts/check_degradation.sh:17]: instead of exiting on first fail consider output every significant changes and then if there are exceeds fail the script

File: .github/workflows/bot.yml [hunk 10]
```diff
@@ -148,3 +148,43 @@ jobs:
...
+          bash .github/scripts/check_degradation.sh
+
+      - name: Comment on PR
+        uses: peter-evans/commit-comment@v3
+        with:
+          msg: |
+            <details>
+            <summary>Benchmark Results</summary>
+            | Test                          | Base         | PR               | % change   |
+            |-------------------------------|--------------|------------------|------------|
```
Comment (meskill) [inline .github/workflows/bot.yml:187]: consider putting this into `critcmp.txt` file itself

File: .github/scripts/check_degradation.sh [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+    before = $3
+    after = $7
+    before_val = ($3 ~ /ns/) ? $3 : ($3 ~ /µs/) ? $3 * 1000 : ($3 ~ /ms/) ? $3 * 1000000 : "invalid"
+    after_val = ($7 ~ /ns/) ? $7 : ($7 ~ /µs/) ? $7 * 1000 : ($7 ~ /ms/) ? $7 * 1000000 : "invalid"
+
+    if (before_val != "invalid" && after_val != "invalid") {
+        change = ((after_val - before_val) / before_val) * 100
+        gsub("%", "", change)  # Remove '%' symbol
+
+        printf "| %-30s | %-20s | %-20s | %-10.2f |\n", item, before, after, change >> "benches/critcmp.txt"
```
Comment (meskill) [inline .github/scripts/check_degradation.sh:13]: the file `critcmp.txt` should be cleared when script start in case it's already present

File: .github/workflows/bot.yml [hunk 12]
```diff
@@ -148,3 +148,43 @@ jobs:
...
+        run: cargo install critcmp
+
+      - name: Run Criterion on branch
+        run: |
+          cargo bench --features criterion -- --save-baseline new_branch
+
+      - name: Fetch target branch
+        run: |
+          git fetch
+          git checkout ${{ github.sha }}
```
Comment (meskill) [inline .github/workflows/bot.yml:170]: `github.sha` points out to the commit that triggered the job i.e. it's the same branch that was checkouted before

File: .github/workflows/bot.yml [hunk 13]
```diff
@@ -148,3 +148,43 @@ jobs:
         uses: stefanzweifel/git-auto-commit-action@v5
         with:
           commit_message: "style: apply lint fixes"
+
+  Check_Degdratation:
+    name: Criterion Benchmarks
```
Comment (meskill) [inline .github/workflows/bot.yml:153]: There is no iai-callgrind benchmarks anymore?

Comment (alankritdabral) [reply .github/workflows/bot.yml:153]: i am not running iai-callgrind as criterion is better for regression checks , but i have kept iai-callgrind files as we might need to see small change in performances in future

File: .github/scripts/check_degradation.sh (hunk 11 above)
Comment (alankritdabral) [reply .github/scripts/check_degradation.sh:13]: yeah will do this

File: .github/workflows/bot.yml (hunk 12 above)
Comment (alankritdabral) [reply .github/workflows/bot.yml:170]: i get it will make changes accordingly

File: .github/scripts/check_degradation.sh (hunk 9 above)
Comment (alankritdabral) [reply .github/scripts/check_degradation.sh:17]: will make the changes accordingly

File: .github/scripts/check_degradation.sh (hunk 8 above)
Comment (alankritdabral) [reply .github/scripts/check_degradation.sh:7]: thanks for pointing it out i never noticed this

File: benches/iai-callgrind/impl_path_string_for_evaluation_context_iai-callgrind.rs (hunk 6 above)
Comment (alankritdabral) [reply benches/iai-callgrind/impl_path_string_for_evaluation_context_iai-callgrind.rs:1]: no it was working but that was too much change in the code might be done in a separate pr

File: .github/scripts/check_degradation.sh (hunk 8 above)
Comment (alankritdabral) [reply .github/scripts/check_degradation.sh:7]: @meskill by calculating the percentage change will give us more accurate results .
as the minimum change in percentage in the above is 1 . What do you think

File: .github/workflows/bot.yml (hunk 13 above)
Comment (meskill) [reply .github/workflows/bot.yml:153]: well, there is no better runner, they are just different.

we still want to have both kind of benchmarks that shows different representation of performance and the way to see the performance regression in both terms. See the [comment](https://github.com/tailcallhq/tailcall/issues/436#issuecomment-1794164324) from the original issue.

As I remember you already had implementation for this and it just was named mistakenly `Build time`, so we can bring it back probably.

To summarize, we want:
- criterion benchmarks
- iai-callgrind benchmarks

both as separate ci job and different script to handle

File: .github/workflows/bot.yml [hunk 14]
```diff
@@ -148,3 +148,58 @@ jobs:
...
+        run: cargo install critcmp
+
+      - name: Run Criterion on branch
+        run: |
+          cargo bench --features criterion -- --save-baseline new_branch
+
+      - name: Switch to main
+        run: |
+          git fetch
+          git checkout main
```
[the remark this answers is not in the document]
Comment (alankritdabral) [reply .github/workflows/bot.yml]: i tried it but its not working  can you show me how its done

File: .github/workflows/bot.yml [hunk 15]
```diff
@@ -148,3 +148,58 @@ jobs:
...
+             ${fileContent}
+
+             </details>
+             `;
+            github.rest.issues.createComment({
+              issue_number: issue_number,
+              owner: owner,
+              repo: repo,
+              body: comment
+            });
```
[the remark this answers is not in the document]
Comment (alankritdabral) [reply .github/workflows/bot.yml]: done

File: .github/workflows/bot.yml [hunk 16]
```diff
@@ -156,3 +156,104 @@ jobs:
         uses: stefanzweifel/git-auto-commit-action@v5
         with:
           commit_message: "style: apply lint fixes"
+
+  Criterion:
+    name: Criterion Benchmarks
+    if: contains(github.event.pull_request.labels.*.name, 'benchmark')
+    runs-on: ubuntu-latest
```
Comment (tusharmath) [inline .github/workflows/bot.yml:163]:
```suggestion
    runs-on: benchmarking-runner
```

File: benches/iai-callgrind/benchmarks.cfg [hunk 17]
```diff
@@ -0,0 +1,4 @@
+json_like_bench_iai_callgrind::batched_body::benchmark_batched_body
+data_loader_bench_iai_callgrind::data_loader::benchmark_data_loader
+impl_path_string_for_evaluation_context_iai_callgrind::bench::bench_main
+request_template_bench_iai_callgrind::bench_to_request::benchmark_to_request
```
Comment (tusharmath) [inline benches/iai-callgrind/benchmarks.cfg:4]: We should not have to update this file everytime we add a benchmark. We are anyways updating `cargo.toml`. This should not be used or autogenerated somehow.

File: .github/contributing.md [hunk 18]
```diff
@@ -47,6 +47,51 @@ Thank you for considering contributing to **Tailcall**! This document outlines t
    cargo test
    ```
 
+## Benchmarks Comparison
+
+**Critrion Benchmarks**
```
Comment (meskill) [inline .github/contributing.md:52]:
```suggestion
### Criterion Benchmarks
```

File: .github/contributing.md [hunk 19]
```diff
@@ -47,6 +47,51 @@ Thank you for considering contributing to **Tailcall**! This document outlines t
...
+   bash .github/scripts/criterion_benchmark.sh
+
+   ```
+
+4. **Check the Results:** If the benchmarks show more than 10% degradation, the script will exit with an error. Please check "benches/critcmp.txt" file to identify the benchmarks that failed and investigate the code changes that might have caused the degradation.
+   The format for the "benches/critcmp.txt" file is:
+   | Test | Base | PR | % change |
+   |-------------------------------|--------------|------------------|------------|
+
+   **Iai-Callgrind Benchmarks**
```
Comment (meskill) [inline .github/contributing.md:76]:
```suggestion
### Iai-Callgrind Benchmarks
```

File: .github/workflows/bot.yml [hunk 20]
```diff
@@ -156,3 +156,104 @@ jobs:
...
+  Iai_callgrind:
+    name: iai_callgrind_benchmarks
+    if: contains(github.event.pull_request.labels.*.name, 'benchmark')
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+
+      - name: Install iai-callgrind-runner
+        run: cargo install --version 0.7.3 iai-callgrind-runner
```
Comment (meskill) [inline .github/workflows/bot.yml:217]: new version of iai-callgrind is out see https://github.com/iai-callgrind/iai-callgrind/blob/main/CHANGELOG.md .

There is now option to fail on bench regression and support for machine output, so probably this pr could be simplified with its update

**Review by meskill — CHANGES_REQUESTED** [review]
